runtime/src/main/java/org/corfudb/runtime/view/stream/AddressMapStreamView.java

```diff
@@ -62,57 +62,61 @@ public AddressMapStreamView(final CorfuRuntime runtime,
 
     @Override
     protected ILogData removeFromQueue(NavigableSet<Long> queue) {
-        // Because the discovery mechanism implemented by this class
-        // does not require to read log entries in advance (requests
-        // full address map in a single call), entries can be read in
-        // batches whenever we have a cache miss. This allows next reads
-        // to be serviced immediately, rather than reading one entry at a time.
-        Long thisRead = queue.pollFirst();
+        boolean readNext;
+        ILogData ld = null;
+        Long currentRead;
 
-        if (thisRead != null) {
-            // In the case that the immediately requested address (thisRead) is not available,
-            // read it along with a batch of subsequent addresses.
+        do {
+            currentRead = queue.pollFirst();
 
-            ILogData ld;
+            if (currentRead == null) {
+                // no more entries, empty queue.
+                return null;
+            }
 
             try {
-                // The read range has a special behaviour, it attempts to read the address of interest,
-                // and in the case it is not in the cache, it reads a batch of addresses
-                // that will be further required, returning only the data for the address of interest.
-
+                // Because the discovery mechanism implemented by this class
+                // does not require to read log entries in advance (only requests
+                // the stream's full address map, without reading the actual data), entries can be read in
+                // batches whenever we have a cache miss. This allows next reads
+                // to be serviced immediately, rather than reading one entry at a time.
+                ld = read(currentRead, queue);
+
+                if (queue == getCurrentContext().readQueue && ld != null) {
+                    // Validate that the data entry belongs to this stream, otherwise, skip.
+                    // This verification protects from sequencer regression (tokens assigned in an older epoch
+                    // that were never written to, and reassigned on a newer epoch)
+                    if (ld.containsStream(this.id)) {
+                        addToResolvedQueue(getCurrentContext(), currentRead, ld);
+                        readNext = false;
+                    } else {
+                        log.trace("getNextEntry[{}]: the data for address {} does not belong to this stream. Skip.",
+                                this, currentRead);
+                        readNext = true;
```

Comment:
Just curious, where do we remove entries that do not belong to the current stream from the queue? So if we hit an entry that does not belong to the queue, won't we still loop forever?